samza-core/src/main/java/org/apache/samza/control/WatermarkManager.java

```diff
@@ -0,0 +1,191 @@
+  private Map<SystemStream, Integer> upstreamTaskCounts;
+
+  public WatermarkManager(String taskName,
+      Multimap<SystemStream, String> inputToTasks,
+      Set<SystemStreamPartition> ssps,
+      StreamMetadataCache metadataCache,
+      MessageCollector collector) {
+    this.taskName = taskName;
+    this.watermarkPerStream = new HashMap<>();
+    this.inputToTasks = inputToTasks;
+    this.metadataCache = metadataCache;
+    this.collector = collector;
+
+    Map<SystemStreamPartition, WatermarkState> states = new HashMap<>();
+    ssps.forEach(ssp -> {
+        states.put(ssp, new WatermarkState());
+        watermarkPerStream.put(ssp.getSystemStream(), TIME_NOT_EXIST);
+      });
+    this.watermarkStates = Collections.unmodifiableMap(states);
+  }
+
+  public void init(IOGraph ioGraph) {
+    this.upstreamTaskCounts = ControlMessageUtils.calculateUpstreamTaskCounts(inputToTasks, ioGraph);
+  }
+
+  /**
+   * Update the watermark based on the incoming watermark message. The message contains
+   * a timestamp and the upstream producer task. The aggregation result is the minimal value
+   * of all watermarks for the stream:
+   * <ul>
+   *   <li>Watermark(ssp) = min { Watermark(task) | task is upstream producer and the count equals total expected tasks } </li>
+   *   <li>Watermark(stream) = min { Watermark(ssp) | ssp is a partition of stream that assigns to this task } </li>
+   * </ul>
+   *
+   * @param envelope the envelope contains {@link WatermarkMessage}
+   * @return watermark envelope if there is a new aggregate watermark for the stream
+   */
+  public Watermark update(IncomingMessageEnvelope envelope) {
+    SystemStreamPartition ssp = envelope.getSystemStreamPartition();
+    WatermarkState state = watermarkStates.get(ssp);
+    WatermarkMessage message = (WatermarkMessage) envelope.getMessage();
+    state.update(message.getTimestamp(), message.getTaskName(), message.getTaskCount());
+
+    if (state.getWatermarkTime() != TIME_NOT_EXIST) {
+      long minTimestamp = watermarkStates.entrySet().stream()
+          .filter(entry -> entry.getKey().getSystemStream().equals(ssp.getSystemStream()))
+          .map(entry -> entry.getValue().getWatermarkTime())
+          .min(Long::compare)
+          .get();
+      Long curWatermark = watermarkPerStream.get(ssp.getSystemStream());
+      if (curWatermark == null || curWatermark < minTimestamp) {
+        watermarkPerStream.put(ssp.getSystemStream(), minTimestamp);
+        return new WatermarkImpl(minTimestamp);
+      }
+    }
+
+    return null;
+  }
+
+  /* package private */
+  long getWatermarkTime(SystemStreamPartition ssp) {
+    return watermarkStates.get(ssp).getWatermarkTime();
+  }
+
+  /**
+   * Send the watermark message to all partitions of an intermediate stream
+   * @param timestamp watermark timestamp
+   * @param systemStream intermediate stream
+   */
+  void sendWatermark(long timestamp, SystemStream systemStream, int taskCount) {
```

Comment:
Yes, I am focusing on the aggregation and propagation for now. Watermark emit from user and high-level API support will be future tasks.